File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/map/submodelelement/operation/OperationVariable.java
package org.eclipse.basyx.submodel.metamodel.map.submodelelement.operation;

import java.util.Map;

import org.eclipse.basyx.submodel.metamodel.api.submodelelement.ISubmodelElement;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.operation.IOperationVariable;
import org.eclipse.basyx.submodel.metamodel.facade.submodelelement.SubmodelElementFacadeFactory;
import org.eclipse.basyx.submodel.metamodel.map.modeltype.ModelType;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.SubmodelElement;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.dataelement.property.Property;
import org.eclipse.basyx.vab.model.VABModelMap;

/**
 * OperationVariable as described by DAAS document An operation variable is a
 * submodel element that is used as input or output variable of an operation.
 * 
 * @author schnicke
 *
 */
public class OperationVariable extends VABModelMap<Object> implements IOperationVariable {
	public static final String MODELTYPE = "OperationVariable";

	/**
	 * 
	 * @param value
	 *            Describes the needed argument for an operation via a submodel
	 *            element of kind=Type
	 */
	public OperationVariable(SubmodelElement value) {
		// Add model type
		putAll(new ModelType(MODELTYPE));
		
		put(Property.VALUE, value);
	}

	public OperationVariable() {
		// Add model type
		putAll(new ModelType(MODELTYPE));
	}
	
	/**
	 * Creates an OperationVariable object from a map
	 * 
	 * @param obj an OperationVariable object as raw map
	 * @return an OperationVariable object, that behaves like a facade for the given map
	 */
	public static OperationVariable createAsFacade(Map<String, Object> obj) {
		if (obj == null) {
			return null;
		}
		
		OperationVariable facade = new OperationVariable();
		facade.setMap(obj);
		return facade;
	}

	public void setValue(ISubmodelElement value) {
		put(Property.VALUE, value);
	}

	@SuppressWarnings("unchecked")
	@Override
	public ISubmodelElement getValue() {
		return SubmodelElementFacadeFactory.createSubmodelElement((Map<String, Object>) get(Property.VALUE));
	}

}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/map/submodelelement/operation/OperationVariable.cpp
/*
 * ModelType.cpp
 *
 *      Author: wendel
 */

#include <BaSyx/submodel/map/submodelelement/operation/OperationVariable.h>

namespace basyx {
namespace submodel {

OperationVariable::OperationVariable()
	: ModelType(IOperationVariable::Path::ModelType)
  , vab::ElementMap{}
{}

OperationVariable::OperationVariable(basyx::object object)
	: vab::ElementMap{object}
{}

std::shared_ptr<ISubmodelElement> OperationVariable::getValue() const
{
	return std::make_shared<SubmodelElement>(this->map.getProperty(IOperationVariable::Path::Value));
}

std::string OperationVariable::getType() const
{
	return this->map.getProperty(Path::Type).GetStringContent();
}

void OperationVariable::setValue(const SubmodelElement & value)
{
	this->map.insertKey(IOperationVariable::Path::Value, value.getMap(), true);
}

void OperationVariable::setValue(const ISubmodelElement & value)
{
  auto map = SubmodelElement(value).getMap();
  this->map.insertKey(IOperationVariable::Path::Value, map, true);
}

void OperationVariable::setType(const std::string & string)
{
	this->map.insertKey(Path::Type, string, true);
}

}
}
